torch/nn/modules/module.py

```diff
@@ -2971,14 +2975,12 @@ def __dir__(self):
 
     def _replicate_for_data_parallel(self):
         replica = self.__new__(type(self))
-        replica.__dict__ = self.__dict__.copy()
+        replica.__dict__ = {k: (v.copy() if isinstance(v, dict) else v) for k, v in self.__dict__.items()}
 
         # replicas do not have parameters themselves, the replicas reference the original
         # module.
-        replica._parameters = {}
-        replica._buffers = replica._buffers.copy()
-        replica._modules = replica._modules.copy()
```

Comment:
@wconstab the rationale is that instead of special casing the decoupling of these `dict` attributes between replicas, we can do this automatically for every `dict` attribute.

Reply:
but, i think the space of things people could add on top of nn.module (e.g. in a subclass) is unbounded, and we can't know if its better to change the behavior for all dicts.  

@albanD might have more insight on this. But IMO it is not clear if changing this is OK

Reply:
This is indeed going to change the behavior of things significantly. For example new hooks being registered after data_parallel, any user dict, etc.

However, since this only seems to affect data_parallel code we're happy to let you decide on this one